### PR TITLE
CRUD for users added

### PR DESCRIPTION
**Pull request by 96mickey**
Server files added.

### CODE REVIEW

File: app/user/userModel.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  email: string = "";
+  phoneNumber: number = 0;
+  address: string = "";
+  role?: number = 0;
+
+  constructor(init?: Partial<UserModel>) {
+    Object.assign(this, init);
+  }
+
+  save() {
```
Comment (samarpan-b) [inline app/user/userModel.ts]: why this ?

File: app/user/userControlModel.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import { UserItem } from "../types";
+import fs from "fs";
+import { getUsersFromFile, p } from "./utils";
+
+export class UserControlModel {
```
Comment (samarpan-b) [inline app/user/userControlModel.ts]: rename this to data-functions

File: app/types.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  firstName: string;
+  middleName: string;
+  lastName: string;
+  email: string;
+  phoneNumber: number;
+  address: string;
+  role?: number;
+}
+
+export interface UserValidation {
```
Comment (samarpan-b) [inline app/types.ts]: why is this copied ?

File: app/types.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+export interface UserItem {
+  id?: string;
+  firstName: string;
+  middleName: string;
+  lastName: string;
+  email: string;
+  phoneNumber: number;
+  address: string;
+  role?: number;
```
Comment (samarpan-b) [inline app/types.ts]: use enum

File: app/user/userModel.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import fs from "fs";
+import { getUsersFromFile, p } from "./utils";
+import { UserItem } from "../types";
+
+export class UserModel {
```
Comment (samarpan-b) [inline app/user/userModel.ts]: use either model or interface

File: app/user/controllers.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+export class Controllers {
+  getAllUsers = (req: Request, res: Response) => {
+    UserMdl.find((users: UserItem) => {
+      res.send({
+        users: users
+      });
+    });
+  };
+
+  getSingleUser = (req: Request, res: Response) => {
```
Comment (samarpan-b) [inline app/user/controllers.ts]: use void return type

File: app/user/utils.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import fs from "fs";
+const path = require("path");
+
+export const p = path.join(
+  path.dirname((<any>process.mainModule).filename),
+  "../data",
+  "user.json"
+);
+
+export const getUsersFromFile = (cb: any) => {
```
Comment (samarpan-b) [inline app/user/utils.ts]: make this more generic

File: app/user/utils.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    if (err) {
+      console.log("err", err);
+      cb([]);
+    } else {
+      cb(JSON.parse(fileContent as any));
+    }
+  });
+};
+
+export const ValidateReq = (
```
Comment (samarpan-b) [inline app/user/utils.ts]: make this more generic

**Review by samarpan-b — CHANGES_REQUESTED** [review]

File: app/user/userModel.ts (hunk 1 above)
Comment (96mickey) [reply app/user/userModel.ts]: this is to push data in json file.

File: app/utils.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+export const pathfunction = (file: string) => {
+  console.log("main module", process.mainModule);
+  return path.join(
+    path.dirname((<any>process.mainModule).filename),
+    "../data",
+    file
+  );
+};
+
+export const getUsersFromFile = (path: any, cb: any) => {
```
Comment (samarpan-b) [inline app/utils.ts]: getDataFromFile

File: app/user/model.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+  constructor(init?: Partial<UserModel>) {
+    Object.assign(this, init);
+  }
+
+  save() {
+    getUsersFromFile("user.json", (users: UserModel[]) => {
+      this.id = Date.now().toString();
+      users.push(this);
+      fs.writeFile(pathfunction("user.json"), JSON.stringify(users), err => {
```
Comment (samarpan-b) [inline app/user/model.ts]: move this to utils

**Review by samarpan-b — CHANGES_REQUESTED** [review]

File: app/user/validation-schema.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    minLength: 5,
+    maxLength: 50
+  };
+  lastName = {
+    type: "string",
+    required: true,
+    minLength: 5,
+    maxLength: 50
+  };
+  email = {
```
Comment (samarpan-b) [inline app/user/validation-schema.ts:20]: add type as string. type is a mandatory attribute

File: app/utils.ts [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+  });
+};
+
+export const writeData = (file: string, data: any[]) => {
+  fs.writeFile(pathfunction(file), JSON.stringify(data), err => {
+    console.log(err);
+  });
+};
+
+export const ValidateReq = (
```
Comment (samarpan-b) [inline app/utils.ts]: if this is user specific move it inside user folder

File: app/utils.ts [hunk 13]
```diff
@@ -0,0 +1,58 @@
+import fs from "fs";
```
Comment (samarpan-b) [inline app/utils.ts]: Move this file to core/utils

**Review by samarpan-b — CHANGES_REQUESTED** [review]
Add comments to each function and class in the code

File: app/user/controller.ts [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    DataFunctions.findOne(req.params.id, (user: User | void) => {
+      if (user) res.json(user);
+      else return res.status(400).send("No such user was found.");
+    });
+  };
+
+  /**
+   * @param {object} req.body is required and should contain data to create a new user. it holds various fields like first name, last name, email etc.
```
Comment (samarpan-b) [inline app/user/controller.ts:43]: fn description missing

**Review by samarpan-b — APPROVED** [review]

**Pull request merged**